### ISSUE TITLE
Proxy unable to connect to Pilot after 1.3 release

### ISSUE THREAD

**Issue by chaudyg** [istio/istio#19996]
**Bug description**

All my pods are stuck with the following error on the `istio-proxy` container on istio 1.3.X and 1.4.X, but not on 1.2.X.
```
Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 0 successful, 0 rejected; ld │
│ s updates: 0 successful, 0 rejected
```

It seems to be related to iptables. When I `exec` into the `istio-proxy` pod, the connection to pilot hangs
```
$ curl https://istio-pilot.istio-system.svc.cluster.local:15010
# HANGING
$ curl -k https://kubernetes.default.svc.cluster.local
# HANGING
```
I *think* the started after the includeInboundPorts changed from the container ports  [in 1.2](https://github.com/istio/istio/blob/1.2.10/install/kubernetes/helm/istio/files/injection-template.yaml#L24) to `*` in [1.3/1.4](https://github.com/istio/istio/blob/1.4.2/install/kubernetes/helm/istio/files/injection-template.yaml#L27)

```
# Release Note
Improved container ports are no longer required in the pod spec. All ports are captured by default.
```

The only way I found to bypass the issues is to set `traffic.sidecar.istio.io/includeInboundPorts: '<list of container ports>'` annotation on my pod.

I am suspecting the problem happens in this environment because I am using Lyft CNI in AWS. Not quite sure how to debug further. Could you help?

Also:
- All the istio components are up and running
- The only component reporting errors are the sidecar proxies (`Envoy proxy is NOT ready`)

**Expected behavior**
Envoy sidecar should be able to connect to Pilot without having to set  `traffic.sidecar.istio.io/includeInboundPorts ` annotation.

**Steps to reproduce the bug**
Here is my setup:
- kops 1.14 using Lyft CNI in AWS.
- kubernetes 1.14.9
- istio: tried  1.4.2 and 1.3.9
- helm installation

**How was Istio installed?**
helm, no configuration options.

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS linux AMI 2 in AWS

**Comment by chaudyg** [comment]
For reference, those are the iptables rules
```
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
-N ISTIO_INBOUND
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-N ISTIO_REDIRECT
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A POSTROUTING -o eth0 -m comment --comment "kube-proxy SNAT" -j MASQUERADE
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
```

**Comment by chaudyg** [comment]
Seems that I am also able to make things work with
```
traffic.sidecar.istio.io/excludeInboundPorts: 15010,15011
```

**Comment by howardjohn** [comment]
iptables shouldn't be intercepting things from user 1337 which is what the proxy will run as. Are you running with a different user somehow? You can see the rules:
```
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
```

**Comment by chaudyg** [comment]
No. I am not running a different user.
```
istio-proxy@infrabin-54c8464ff8-7cq52:/$ ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
istio-p+     1     0  0 Jan08 ?        00:01:56 /usr/local/bin/pilot-agent proxy sidecar --domain infrabin.svc.cluster.local --configPath /etc/istio/proxy --binaryPath /usr/local/bi
istio-p+    20     1  0 Jan08 ?        00:01:17 /usr/local/bin/envoy -c /etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cl
istio-p+    47     0  0 09:21 ?        00:00:00 bash
istio-p+    61    47  0 09:22 ?        00:00:00 ps -ef
istio-proxy@infrabin-54c8464ff8-7cq52:/$ stat -c "%u %g" /proc/1/
1337 1337
istio-proxy@infrabin-54c8464ff8-7cq52:/$ stat -c "%u %g" /proc/20/
1337 1337
```

Could be related to the fact i am using a ENI CNI (lyft)? see this [description](https://github.com/lyft/cni-ipvlan-vpc-k8s#internet-egress)

```
$ ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0@veth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UNKNOWN mode DEFAULT group default
    link/ether 0a:f4:c4:82:62:d4 brd ff:ff:ff:ff:ff:ff
4: veth0@if162: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP mode DEFAULT group default
    link/ether 0e:6c:e0:b7:88:70 brd ff:ff:ff:ff:ff:ff link-netnsid 0
$ ip route
default via 172.16.228.164 dev veth0
172.16.0.0/19 dev eth0 proto kernel scope link src 172.16.14.181
172.16.0.0/16 via 172.16.0.1 dev eth0
172.16.228.164 dev veth0 scope link
172.17.0.0/16 via 172.16.0.1 dev eth0
172.18.0.0/16 via 172.16.0.1 dev eth0
172.20.0.0/16 via 172.16.0.1 dev eth0
```

**Comment by howardjohn** [comment]
hm. I am not familiar with that CNI but it seems like for sure traffic is blocked.

What I am really confused is why `traffic.sidecar.istio.io/excludeInboundPorts: 15010,15011`. That is supposed to be for inbound traffic, but you are trying to open up outbound traffic.. Given that makes a difference though its pretty clear something funky is going on with the interaction between iptables and your CNI?

**Comment by chaudyg** [comment]
After a long investigation, finally started to understand how the CNI works. Here is the flow for envoy making a request to Pilot using it's service VIP.
```
Envoy ---> veth0 (pod netns)  ---> vethX (host netns) --> veth0 (pod netns) --> eth0 (pod netns) 
```

In the pod network namespace, they are 2 network interfaces:
- eth0 : all the traffic to the VPC. This includes any requests to pods
- veth0: anything else. This includes internet traffic but also Kube Proxy CIDR.

When envoy makes a request to pilot, the requests leaves the pod network namespace via veth0 and lands on the host network namespace.

Then, kube-proxy Iptables rules does it's magic. The destination IP dNAT to one of pilot pod IPs.

Then, the packet is sent back into the pod. There is an `ip rule/route` on the host network namespace that sends the packet back into the pod. This is done to ensure all the traffic from to the VPC exits via the same interface.

Sadly, When the packet inbound back into veth0, it gets trapped into  the `ISTIO_IN_REDIRECT` iptables rule.

Long story short. This will affect anyone using the AWS lyft ENI/CNI . Maybe I should add an iptable rule that says `ignore all the traffic coming from veth0` but I am not sure it's the right way forward.  @howardjohn what would you suggest?

```-A ISTIO_INBOUND -p tcp -m tcp -i veth0 -j RETURN```

**Comment by howardjohn** [comment]
@rlenglet or @lambdai can help more here, this is a lot deeper than my iptables knowledge :slightly_smiling_face:

**Comment by rlenglet** [comment]
Note that Istio doesn't touch the `POSTROUTING` chain. I notice that this pod has this rule, which was not created by Istio:
```
-A POSTROUTING -o eth0 -m comment --comment "kube-proxy SNAT" -j MASQUERADE
```

**Comment by rlenglet** [comment]
@chaudyg can you please post the routing table inside the pod?

**Comment by lambdai** [comment]
The output of `ip link show` is interesting. 
Usually we shouldn't see the other end of veth0
What's the output of `ip address`?

**Comment by rlenglet** [comment]
> Usually we shouldn't see the other end of veth0

I don't think this is the other end of a veth pair. This is another interface created via ENI.

**Comment by rlenglet** [comment]
> Sadly, When the packet inbound back into veth0, it gets trapped into the ISTIO_IN_REDIRECT iptables rule.

That shouldn't be an issue. That rule is normally not hit by return traffic.

So just to confirm my understanding, `eth0` is the "IPvlan interface" and `veth0` is the  "unnumbered point-to-point interface".

I suspect that the `MASQUERADE` rule is messing up with the Istio rules:
```
-A POSTROUTING -o eth0 -m comment --comment "kube-proxy SNAT" -j MASQUERADE
```
I don't understand why that rule is created by the CNI plugin, since the pod has only one IP address in this setup.

**Comment by chaudyg** [comment]
@rlenglet @lambdai  @howardjohn  Thank you for your feedback on this.


### The Lyft CNI
First, here is my understanding of how the Lyft CNI works with Envoy. They are 3 important scenarios
1. Pod to VPC (or pod to pod)
2. Pod to service VIP
3. Pod to internet

#### Pod to VPC
*this works*
![image](https://user-images.githubusercontent.com/7817963/73729559-2cdcf500-472d-11ea-8ce6-e753fa0016fa.png)

A tcp is sent by an application inside the pod. The destination interface is eth0 (all the traffic to the VPC)
```
ip route get 172.16.68.1
172.16.68.1 dev eth0  src 172.16.68.53
```

1. The following iptables rules in the pod netns redirect the request to envoy.
```
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A POSTROUTING -o eth0 -m comment --comment "kube-proxy SNAT" -j MASQUERADE
```

2. Envoy does it's magic and sends the request to the pod IP
The following iptables rules match:
```
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
```

3. The requests exits the pod netns via eth0 and the root netns via eth1 (ENI)

#### Pod to internet
*This works*
This is the default route for any request not going to the VPC of kube-proxy CIDR
![image](https://user-images.githubusercontent.com/7817963/73730839-6878be80-472f-11ea-967b-ab96b497b49c.png)

```
ip route get 13.248.155.102
13.248.155.102 via 172.16.230.147 dev veth0  src 172.16.68.53
```
1. Iptables rules redirect the request to envoy (same as above).
2. Envoy does it's magic and sends the request to destination IP
3. The packets arrive in the root netns via vethX (one of those per pod) and is forwarded to eth0 (default route)
4. The pakcets leave the host via eth0

#### Pod to service VIP
*this does not work*
The application (or envoy) make a request to a service VIP (ie pilot)
![image](https://user-images.githubusercontent.com/7817963/73732384-fd7cb700-4731-11ea-8a53-05698339d995.png)

```
ip route get 10.254.1.2
10.254.1.2 via 172.16.230.147 dev veth0  src 172.16.68.53
```
1. Iptables rules redirect the request to envoy (same as above).
2. Envoy does it's magic and sends the request to destination VIP (same as before) via veth0 (pod netns)->vethX (root netns)
3. kube proxy iptables DNAT to the correct POD ip
4. On the root netns, the folllowing ip route rules send the request back into the Pod netns
```
$ ip rule
[...]
1024:	from all iif vethf6d62357 lookup 306

$ ip route show table 306
[...]
172.16.0.0/16 via 172.16.68.53 dev vethf6d62357
```

5. This is where everything breaks.
The following iptables match
```
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006 🤕
```
So, when envoy tries to fetch its configuration from pilot, the packet gets trapped into envoy.

### The patch I tried
I've added the following iptables rule to fix it in my cluster

```
-A ISTIO_INBOUND -p tcp -m tcp -i veth0 -j RETURN
```
The effect is:
![image](https://user-images.githubusercontent.com/7817963/73732778-acb98e00-4732-11ea-8665-47e159844ea2.png)

1-4. Same as before
5. The following iptables rule match
```
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp -m tcp -i veth0 -j RETURN
-A POSTROUTING -o eth0 -m comment --comment "kube-proxy SNAT" -j MASQUERADE
```
Note the masquerade rule used here.

6.  The packets leave via eth0->eth1 to the VPC.

### Where to go from here?
I found the following flag `istio-iptables.sh` 
```
  echo '  -k: Comma separated list of virtual interfaces whose inbound traffic (from VM)'
  echo '      will be treated as outbound (optional)'
```  
https://github.com/istio/cni/blob/1.4.3/tools/packaging/common/istio-iptables.sh#L52-L53

When I pass `-k veth0` to the init script, the following 2 iptables rules are added. The second one seems to be the one I need.  I am not sure why the first one is used for. But both conflicts with one another. 

```
-A PREROUTING -i veth0 -j ISTIO_REDIRECT 
-A PREROUTING -i veth0 -j RETURN
```

I hope that helps. What do you think?

**Comment by chaudyg** [comment]
@rlenglet @lambdai @howardjohn 

How would you feel if I raise a pr remove the Istio redirect rule for a given interface when using the `-k` flag?

See:
```
-A PREROUTING -i veth0 -j ISTIO_REDIRECT 
-A PREROUTING -i veth0 -j RETURN
```
